serviceregistration/kubernetes/retry_handler.go

```diff
@@ -0,0 +1,112 @@
+package kubernetes
+
+import (
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/vault/serviceregistration/kubernetes/client"
+)
+
+// How often to retry sending a state update if it fails.
+var retryFreq = 5 * time.Second
+
+// retryHandler executes retries.
+// It is thread-safe.
+type retryHandler struct {
+	// These don't need a mutex because they're never mutated.
+	logger             hclog.Logger
+	namespace, podName string
+	client             *client.Client
+
+	// This gets mutated on multiple threads.
+	patchesToRetry     []*client.Patch
+	patchesToRetryLock sync.Mutex
+}
+
+// Run runs at an interval, checking if anything has failed and if so,
+// attempting to send them again.
+func (r *retryHandler) Run(shutdownCh <-chan struct{}, wait *sync.WaitGroup) {
+	// Make sure Vault will give us time to finish up here.
+	wait.Add(1)
+	defer wait.Done()
+
+	retry := time.NewTicker(retryFreq)
+	for {
+		select {
+		case <-shutdownCh:
+			return
+		case <-retry.C:
+			r.retry()
+		}
+	}
+}
+
+// Add adds a patch to be retried until it's either completed without
+// error, or no longer needed.
+func (r *retryHandler) Add(patch *client.Patch) error {
+	r.patchesToRetryLock.Lock()
+	defer r.patchesToRetryLock.Unlock()
+
+	// - If the patch is a dupe, don't add it.
+	// - If the patch reverts another, remove them both.
```

Comment:
Is this correct?  If my current state is "active = true", then applying patches "active = true", "active = false" should result in "active = false", but in this case they're both removed so things remain unchanged.

More generally, is it the retryHandler's responsibility to adjust the set of patches at all? The function description sounds correct:  add a patch to a list of patches to be applied, and the retryer will attempt to make that happen. As is, depending of which "batch" a patch ends up in might result in different behavior because of ignoring and combining. And if a user explicitly wants to patch "a=42" followed by "a=42", should that be prevented?

Reply:
Well, the retryHandler is only equipped to handler booleans at the moment, and errors if someone sends in anything else. 

Presently there are only 8 operations that can possibly be sent to it, boiling down to:

1. vault-active = true
2. vault-active = false
3. vault-sealed = true
4. vault-sealed = false
5. vault-perf-standby = true
6. vault-perf-standby = false
7. vault-initialized = true
8. vault-initialized = false

So I'm imagining a scenario where we get a notification for 1 and we can't deliver it, so the retry handler endlessly tries and tries. But then it's been a while, and we get a notification for 2. That means the other end has returned to the last state it knew about on its own, so we don't need to continue attempting to send either 1 or 2.

At most, there can only ever be 4 patches that are being retried, because there are only 4 dynamic labels that can be out of sync at a time.

I hope that makes sense, I know the code is a bit strange. Are you thinking that instead of cancelling them out, we should simply both notify 1, then notify 2?

Reply:
Just capturing our offline discussion: using a map to maintain the latest requested state may be simpler.

Reply:
@kalafut just updated this to use a last-wins map per our conversation: https://github.com/hashicorp/vault/pull/8249/commits/04bc9951b16c4a54c4dfb90f23e53611200b4a73